### PR TITLE
User-defined functors (new types & refactor)

### PR DESCRIPTION
**Pull request by darth-tytus**
- Add transformer pass that moves type information from functor declaration to instances
- Unify interface for functors
- Add float and unsigned support
- Unify functors analysis
- Add support for a 64-bit domain for user-defined functors
- Fix/extend user-defined functors test

### CODE REVIEW

**Comment by b-scholz** [comment]
Can you reformat the code?
```
+ git diff -m
diff --git a/src/AstFunctorDeclaration.h b/src/AstFunctorDeclaration.h
index d81fb65..9317e64 100644
--- a/src/AstFunctorDeclaration.h
+++ b/src/AstFunctorDeclaration.h
@@ -32,9 +32,9 @@ namespace souffle {
 
 class AstFunctorDeclaration : public AstNode {
 public:
-    AstFunctorDeclaration(const std::string& name, const std::vector<RamTypeAttribute>& argsTypes,
-            RamTypeAttribute returnType)
-            : name(name), argsTypes(argsTypes), returnType(returnType) {
+    AstFunctorDeclaration(
+            const std::string& name, std::vector<RamTypeAttribute> argsTypes, RamTypeAttribute returnType)
+            : name(name), argsTypes(std::move(argsTypes)), returnType(returnType) {
         assert(name.length() > 0 && "functor name is empty");
     }
 
```

**Comment by darth-tytus** [comment]
> Can you reformat the code?
> 
> ```
> + git diff -m
> diff --git a/src/AstFunctorDeclaration.h b/src/AstFunctorDeclaration.h
> index d81fb65..9317e64 100644
> --- a/src/AstFunctorDeclaration.h
> +++ b/src/AstFunctorDeclaration.h
> @@ -32,9 +32,9 @@ namespace souffle {
>  
>  class AstFunctorDeclaration : public AstNode {
>  public:
> -    AstFunctorDeclaration(const std::string& name, const std::vector<RamTypeAttribute>& argsTypes,
> -            RamTypeAttribute returnType)
> -            : name(name), argsTypes(argsTypes), returnType(returnType) {
> +    AstFunctorDeclaration(
> +            const std::string& name, std::vector<RamTypeAttribute> argsTypes, RamTypeAttribute returnType)
> +            : name(name), argsTypes(std::move(argsTypes)), returnType(returnType) {
>          assert(name.length() > 0 && "functor name is empty");
>      }
>  
> ```

It seems that clang-format and linter don't agree on the correct formatting. Is this possible?

**Comment by mmcgr** [comment]
clang-format and the linter should agree, but the style check of clang-format matters is the one that matters, so go with that. The linter only objects to the results of the static analysis, not the format. It would be nice if they agreed, though, so I'll try and fix that.

**Comment by darth-tytus** [comment]
> clang-format and the linter should agree, but the style check of clang-format matters is the one that matters, so go with that. The linter only objects to the results of the static analysis, not the format. It would be nice if they agreed, though, so I'll try and fix that.

Apologies, it's my fault, the changed formatting implied by the linter was caused by another change.

**Comment by mmcgr** [comment]
The number of merges in this branch is starting to rival that of the relevant commits. Could you rebase to clean this up a bit?

File: src/AstArgument.h [hunk 1]
```diff
@@ -356,18 +367,43 @@ class AstUserDefinedFunctor : public AstFunctor {
...
+    RamTypeAttribute getReturnType() const override {
+        return returnType;
+    }
+
+    void setArgsTypes(std::vector<RamTypeAttribute> types) {
+        assert(types.size() == args.size() && "Size of types must match size of arguments");
+        argTypes = types;
+    }
+
+    const std::vector<RamTypeAttribute>& getArgsTypes(void) const {
```
Comment (mmcgr) [inline src/AstArgument.h]: We don't normally declare void parameters for functions.

File: src/AstArgument.h [hunk 2]
```diff
@@ -356,18 +367,43 @@ class AstUserDefinedFunctor : public AstFunctor {
...
-        for (auto& cur : args) {
-            res->args.emplace_back(cur->clone());
+        auto res = new AstUserDefinedFunctor(name);
+        // Set args
+        for (auto& arg : args) {
+            res->args.emplace_back(arg->clone());
+        }
+        // Set types
+        // Only copy types if they have already been set.
+        if (argTypes.size() > 0) {
```
Comment (mmcgr) [inline src/AstArgument.h]: Use `!empty()` rather than `size() > 0`:
```suggestion
        if (!argTypes.empty()) {
```

File: src/parser.yy [hunk 3]
```diff
@@ -1534,38 +1532,42 @@ comp_init
...
     }
   | non_empty_functor_arg_type_list[curr_list] COMMA functor_type {
-        $$ = $curr_list + $functor_type;
+        $$ = $curr_list;
+        $$.push_back($functor_type);
+        $curr_list.clear();
     }
   ;
 
+
```
Comment (mmcgr) [inline src/parser.yy]: One blank line for spacing is enough

File: tests/evaluation.at [hunk 4]
```diff
@@ -4,6 +4,8 @@
 # - https://opensource.org/licenses/UPL
 # - <souffle root>/licenses/SOUFFLE-UPL.txt
 
+
+
```
Comment (mmcgr) [inline tests/evaluation.at]: These blank lines aren't needed.

File: tests/evaluation.at [hunk 5]
```diff
@@ -118,3 +120,6 @@ POSITIVE_TEST([unpacking],[evaluation])
 POSITIVE_TEST([unsigned_operations], [evaluation])
 POSITIVE_TEST([unused_constraints],[evaluation])
 POSITIVE_TEST([x9],[evaluation])
+
+
+POSITIVE_FUNCTOR_TEST([functors],[evaluation])
```
Comment (mmcgr) [inline tests/evaluation.at]: I know I'm the one that moved functors from interface to evaluation, but it *is* a form of interface test so perhaps it should be moved back ...

The important change was the addition of accepting the configuration flags.

File: src/AstArgument.h [hunk 6]
```diff
@@ -305,13 +312,17 @@ class AstIntrinsicFunctor : public AstFunctor {
         function = functor;
     }
 
+    bool isOverloaded() const override {
```
Comment (b-scholz) [inline src/AstArgument.h]: Again - we don't like this type of method in an AST class. AST is only there to describe programs.

File: src/AstArgument.h [hunk 7]
```diff
@@ -305,13 +312,17 @@ class AstIntrinsicFunctor : public AstFunctor {
...
+        return isOverloadedFunctor(function);
+    }
+
     /** get the return type of the functor. */
-    RamTypeAttribute getReturnType() const {
+    RamTypeAttribute getReturnType() const override {
```
Comment (b-scholz) [inline src/AstArgument.h]: Can we change the name of the type. A RAM type should not be used in AST.

File: src/AstArgument.h [hunk 8]
```diff
@@ -258,9 +258,16 @@ class AstTerm : public AstArgument {
  */
 
 class AstFunctor : public AstTerm {
+public:
+    virtual bool isOverloaded() const {
```
Comment (b-scholz) [inline src/AstArgument.h]: Can we remove this method from AstFunctor? We don't like methods in the class hierarchy that goes beyond the description of an Ast element. If they provide any semantic information, we would prefer to place them in their own analysis class/util class (depending on whether they need state).

**Pull request merged**